### ISSUE TITLE
potential bug in output of KaSim results?

### ISSUE THREAD

**Issue by yarden** [Kappa-Dev/KaSim#330]
What follows might be my own pedestrian misunderstanding, but I'm getting counter-intuitive results when plotting Kappa simulations and also when running ``KaSim`` directly. Consider this [two-agent kappa program](https://gist.github.com/yarden/6366f0e4d7f2a592f60d368e5fc36611). It contains a perturbation at T=50 that adds 100 units of ``Gal()`` agent:

```
%mod: [T]=50 || [false] do $ADD 100 Gal()
```

``Gal()`` is initially 0, so its value should increase at T=100, yet the plot shows it flat at 100 starting at around T=1:

![kasim_result](https://cloud.githubusercontent.com/assets/96318/22219551/2f66174c-e179-11e6-88c2-c896764e6302.png)

The simulation correctly shows that agents that depend on ``Gal()`` are only made at around T=50, as expected, so it doesn't look like ``Gal()``'s value is actually interpreted internally as being at level 100 starting at T=1.

When I run ``KaSim`` directly, it matches the plotted results:

```
$ KaSim -i toy_model.ka -l 150
+ Self seeding...
+ Command line to rerun is: 'KaSim' '-i' 'toy_model.ka' '-l' '150' -seed 298615321
Parsing toy_model.ka...
...[snip]...
$ head data.csv 
# Output of 'KaSim' '-i' 'toy_model.ka' '-l' '150' -seed 298615321
"[T]", "[T]glu_sensor_var", "[T]glu_sensor_varglu_sensor_active", "[T]glu_sensor_varglu_sensor_activegal"
0., 50, 0, 0
1., 50, 0, 100
2., 50, 0, 100
3., 50, 0, 100
4., 50, 0, 100
5., 50, 0, 100
6., 50, 0, 100
7., 50, 0, 100
```

There's another potential issue with the ``KaSim`` output. The variables I set to observe were:

```
%obs: 'glu_sensor_var' |GluSensor()|
%obs: 'glu_sensor_active' |GluSensor(gal_ext!gal_ext.Gal)|
%obs: 'gal' |Gal()|
```

My expectation was three columns in output, ``[T]glu_sensor_var``, ``[T]glu_sensor_active``, and ``[T]gal``, but it looks like the string ``glu_sensor_var`` got prepended to each variable name. In any case, you can see that the KaSim result matches the plotting: ``gal`` is 0 at T=0, but 100 at T=1 rather than T=100, which is what I can't explain.

Thanks very much, Yarden

**Comment by yarden** [comment]
Does this also fix the column name problem?

**Comment by pirbo** [comment]
Sorry, I forgot to reply about this part: No, it doesn't.
This is issue is probably an OCaml standard library bug! What OCaml version are you using?
OCaml 4.02.1 (if I remember correctly, I'm not sure about the number) has this weird bug that it does not empty its internal buffers between printfs. So, if you try to printf "foo"; printf "bar"; printf "toto", you end printing "foo" "foobar" "foobartoto".
I've never looked seriously for a fix because OCaml 4.02.3 (that were already out at the time I discovered the bug) and beyond are not affected but I can if necessary.

**Comment by yarden** [comment]
I believe I ran KaSim with OCaml 4.02.3:

```
$ ocaml
        OCaml version 4.02.3

# Printf.printf "foo";;
foo- : unit = ()
# Printf.printf "bar";;
bar- : unit = ()
```

**Comment by pirbo** [comment]
I'm sorry, I need more about your configuration.
I cannot reproduce with OCaml 4.02.3 installed by opam on debian amd64. (but we've seen the bug several time already, you're not alone :-))

**Comment by pirbo** [comment]
3 remarks:
- There must a problem somewhere with your OCaml installation because I've found the bug: https://caml.inria.fr/mantis/view.php?id=6824 and it is indeed fixed in OCaml 4.02.3. Anyway, because it was a recurrently raised issue, I've pushed a workaround so that KaSim is not affected anymore.
- `%mod: [T]=50 do $ADD 100 Gal()` is enough (no need for the `|| [false]` part :-)) but you must not have invented it yourself so I'm curious, do you remember from where you copy `[T]=50 || [false]`?
- If you read the warnings produced by KaSim (but they are hidden in the log tab in the UI, we must change that), you'll see (please improve the english)
```
File "...", line 8, characters 16-25:
Warning: Rule induced deletion AND creation of the agent GluSensor
File "...", line 8, characters 35-38:
Warning: Rule induced deletion AND creation of the agent Gal
```
Indeed, because of the reordering in the right hand side
```
'Gal.GluSensor' GluSensor(gal_ext),Gal(gal_ext) -> Gal(gal_ext!1),GluSensor(gal_ext!1) @ 'glu_sense_on_rate'
```
 means destroy the GLuSensor and the Gal and create fresh new ones bound (This is the F### longest prefix convention explained in the manual). To say "bind a Gal and a GluSensor", you must say 
```
'Gal.GluSensor' GluSensor(gal_ext),Gal(gal_ext) -> GluSensor(gal_ext!1),Gal(gal_ext!1) @ 'glu_sense_on_rate'
```
